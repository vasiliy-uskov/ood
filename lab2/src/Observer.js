"use strict";
exports.__esModule = true;
var typescript_map_1 = require("typescript-map");
var Observable = /** @class */ (function () {
    function Observable() {
        this._observers = new typescript_map_1.TSMap();
        this._observersRemoveList = [];
    }
    Observable.prototype.addObserver = function (observer) {
        this._observers.set(observer, true);
    };
    Observable.prototype.removeObserver = function (observer) {
        this._observers.set(observer, false);
        this._observersRemoveList.push(observer);
    };
    Observable.prototype.notifyObservers = function () {
        var observers = this._observers.keys();
        for (var _i = 0, observers_1 = observers; _i < observers_1.length; _i++) {
            var observer = observers_1[_i];
            if (this._observers.get(observer)) {
                observer.update(this._getObservableData());
            }
        }
        ;
        this._clearRemoveList();
    };
    Observable.prototype._clearRemoveList = function () {
        for (var _i = 0, _a = this._observersRemoveList; _i < _a.length; _i++) {
            var observer = _a[_i];
            this._observers["delete"](observer);
        }
        this._observersRemoveList = [];
    };
    return Observable;
}());
exports.Observable = Observable;
