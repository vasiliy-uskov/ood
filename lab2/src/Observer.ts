import { TSMap } from "typescript-map"

interface IObserver<T> {
    update(data: T):void;
}

abstract class Observable<T> {
    public addObserver(observer: IObserver<T>, priority: number = this._greaterPriority) {
        this._greaterPriority = priority > this._greaterPriority ? priority : this._greaterPriority;
        this._observers.set(observer, {
            active: true,
            priority,
        });
    }

    public removeObserver(observer: IObserver<T>) {
        this._observers.get(observer).active = false;
        this._observersRemoveList.push(observer);
    }

    public notifyObservers() {
        const sortedObservers = this._observers.keys().slice();
        sortedObservers.sort((observer1: IObserver<T>, observer2: IObserver<T>) => {
            const priority1 = this._observers.get(observer1).priority;
            const priority2 = this._observers.get(observer2).priority;
            return priority1 > priority2 ? 1 : (priority1 == priority2 ? 0 : -1);
        });
        for (const observer of sortedObservers) {
            if (this._observers.get(observer).active) {
                observer.update(this._getObservableData());
            }
        }
        this._clearRemoveList();
    }

    protected abstract _getObservableData() : T

    private _clearRemoveList() {
        for (const observer of this._observersRemoveList) {
            this._observers.delete(observer);
        }
        this._observersRemoveList = [];
    }

    private _observers: TSMap<IObserver<T>, {active: boolean, priority: number}> = new TSMap();
    private _observersRemoveList: Array<IObserver<T>> = [];
    private _greaterPriority: number = 0;
}

export {IObserver};
export {Observable};