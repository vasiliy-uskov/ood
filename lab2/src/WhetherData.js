"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
exports.__esModule = true;
var Observer_1 = require("./Observer");
var WhetherInfo = /** @class */ (function () {
    function WhetherInfo(temperature, humidity, pressure) {
        this.humidity = humidity;
        this.temperature = temperature;
        this.pressure = pressure;
    }
    return WhetherInfo;
}());
var Statistics = /** @class */ (function () {
    function Statistics() {
        this._minValue = Infinity;
        this._maxValue = -Infinity;
        this._accValue = 0.;
        this._accCount = 0;
    }
    Statistics.prototype.addValue = function (value) {
        if (value < this._minValue) {
            this._minValue = value;
        }
        if (value > this._maxValue) {
            this._maxValue = value;
        }
        this._accValue += value;
        ++this._accCount;
    };
    Statistics.prototype.minValue = function () {
        return this._minValue;
    };
    Statistics.prototype.maxValue = function () {
        return this._minValue;
    };
    Statistics.prototype.average = function () {
        return this._accCount ? this._accValue / this._accCount : 0;
    };
    return Statistics;
}());
var StatisticsDisplay = /** @class */ (function () {
    function StatisticsDisplay() {
        this._temperatureStatistics = new Statistics();
        this._humidityStatistics = new Statistics();
        this._pressureStatistics = new Statistics();
    }
    StatisticsDisplay.prototype.update = function (data) {
        this._temperatureStatistics.addValue(data.temperature);
        this._humidityStatistics.addValue(data.humidity);
        this._pressureStatistics.addValue(data.pressure);
        this.displayStatistics("Temp", this._temperatureStatistics);
        this.displayStatistics("Pressure", this._pressureStatistics);
        this.displayStatistics("Humidity", this._humidityStatistics);
    };
    StatisticsDisplay.prototype.displayStatistics = function (statisticsName, statistics) {
        console.log("Max", statisticsName, statistics.maxValue());
        console.log("Min", statisticsName, statistics.minValue());
        console.log("Average", statisticsName, statistics.average());
        console.log("--------");
    };
    return StatisticsDisplay;
}());
exports.StatisticsDisplay = StatisticsDisplay;
var WhetherData = /** @class */ (function (_super) {
    __extends(WhetherData, _super);
    function WhetherData() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._temperature = 0.;
        _this._humidity = 0.;
        _this._pressure = 750.;
        return _this;
    }
    WhetherData.prototype.temperature = function () {
        return this._temperature;
    };
    WhetherData.prototype.humidity = function () {
        return this._humidity;
    };
    WhetherData.prototype.pressure = function () {
        return this._pressure;
    };
    WhetherData.prototype.measurementsChanged = function () {
        this.notifyObservers();
    };
    WhetherData.prototype.setMeasurements = function (temperature, humidity, pressure) {
        this._humidity = humidity;
        this._temperature = temperature;
        this._pressure = pressure;
        this.measurementsChanged();
    };
    WhetherData.prototype._getObservableData = function () {
        return new WhetherInfo(this.temperature(), this.humidity(), this.pressure());
    };
    return WhetherData;
}(Observer_1.Observable));
exports.WhetherData = WhetherData;
