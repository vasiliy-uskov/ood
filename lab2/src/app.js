"use strict";
exports.__esModule = true;
var WhetherData_1 = require("./WhetherData");
var WhetherData_2 = require("./WhetherData");
var wd = new WhetherData_1.WhetherData();
var display = new WhetherData_2.StatisticsDisplay();
wd.addObserver(display);
wd.setMeasurements(3, 0.7, 760);
wd.setMeasurements(4, 0.8, 761);
wd.removeObserver(display);
wd.setMeasurements(10, 0.8, 761);
wd.setMeasurements(-10, 0.8, 761);
