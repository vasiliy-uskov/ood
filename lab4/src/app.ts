import {Canvas} from "./Canvas";
import {Painter} from "./Pinter";
import {Designer} from "./Designer";
import {ShapeFactory} from "./ShapeFactory";

const canvas = new Canvas(50, 50);
const painter = new Painter();
const designer = new Designer(new ShapeFactory());
painter.drawPicture(designer.createDraft([
    {
        type: "ellipse",
        color: 0,
        center: {x: 1, y: 1},
        radiusX: 1,
        radiusY: 1
    },
    {
        type: "triangle",
        color: 1,
        vertex1: {x: 0, y: 1},
        vertex2: {x: 1, y: 0},
        vertex3: {x: 0, y: 0},
    },
    {
        type: "rectangle",
        color: 1,
        vertex1: {x: 1, y: 1},
        vertex2: {x: 2, y: 2},
    },
    {
        type: "polygon",
        color: 1,
        vertexes: [
            {x: 0, y: 1},
            {x: 2, y: 3},
            {x: 4, y: 5},
            {x: 6, y: 7},
            {x: 8, y: 9},
            {x: 9, y: 8},
            {x: 7, y: 6}
        ]
    }
]), canvas);

console.log(canvas.getPicture());