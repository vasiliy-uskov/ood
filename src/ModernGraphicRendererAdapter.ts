import {ICanvas} from "./graphic_lib";
import {ModernGraphicsRenderer, Point} from "./modern_graphic_lib";

export class ModernGraphicRendererAdapter implements ICanvas {
	constructor(printFn: (str: string) => void) {
		this._modernRenderer = new ModernGraphicsRenderer(printFn);
	}

	setColor(color: number): void {
		const red = Math.floor(color / 256 / 256);
		const green = Math.floor((color - red * 256 * 256) / 256);
		const blue = Math.floor((color - red * 256 * 256 - green * 256));
		this._color = {
			r: red / 255,
			g: green / 255,
			b: blue / 255,
			a: 1,
		}
	}

	moveTo(x: number, y: number): void {
		if (this._lastPoint)
		{
			this._modernRenderer.endDraw();
		}
		this._modernRenderer.beginDraw();
		this._lastPoint = {x, y};
	}

	lineTo(x: number, y: number): void {
		if (!this._lastPoint)
		{
			throw new Error('Can not draw line without start point');
		}
		const newPoint = {x, y};
		this._modernRenderer.drawLine(this._lastPoint, newPoint, this._color);
		this._lastPoint = newPoint;
	}

	close(): void {
		if (this._lastPoint)
		{
			this._modernRenderer.endDraw();
			this._lastPoint = null;
		}
	}

	private _color = {r: 0, g: 0, b: 0, a: 0};
	private _lastPoint: Point | null = null;
	private readonly _modernRenderer = new ModernGraphicsRenderer(console.log);
}