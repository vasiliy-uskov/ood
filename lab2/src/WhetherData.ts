import {Observable} from "./Observer";
import {IObserver} from "./Observer";
import {TSMap} from "typescript-map";

class WhetherInfo {
    constructor(temperature: number, humidity: number, pressure: number, id: string) {
        this.humidity = humidity;
        this.temperature = temperature;
        this.pressure = pressure;
        this._id = id
    }

    id(): string {
        return this._id;
    }

    public _id: string;
    public temperature: number;
    public humidity: number;
    public pressure: number;
}

class Statistics {
    public addValue(value: number) {
        if (value < this._minValue) {
            this._minValue = value
        }
        if (value > this._maxValue) {
            this._maxValue = value
        }
        this._accValue += value;
        ++this._accCount;
    }

    public minValue(): number {
        return this._minValue;
    }

    public maxValue(): number {
        return this._minValue;
    }

    public average(): number {
        return this._accCount ? this._accValue / this._accCount : 0;
    }

    private _minValue: number = Infinity;
    private _maxValue: number = -Infinity;
    private _accValue: number = 0.;
    private _accCount: number = 0;
}

class StatisticsDisplay implements IObserver<WhetherInfo> {
    public update(data: WhetherInfo) {
        this._initStatistic(data.id());
        const temperatureStatistics = this._temperatureStatistics.get(data.id());
        const humidityStatistics = this._humidityStatistics.get(data.id());
        const pressureStatistics = this._pressureStatistics.get(data.id());
        temperatureStatistics.addValue(data.temperature);
        humidityStatistics.addValue(data.humidity);
        pressureStatistics.addValue(data.pressure);
        this.displayStatistics(`Temp ${data.id()}`, temperatureStatistics);
        this.displayStatistics(`Pressure ${data.id()}`, pressureStatistics);
        this.displayStatistics(`Humidity ${data.id()}`, humidityStatistics);
    }

    public displayStatistics(statisticsName: string, statistics: Statistics) {
        console.log("Max", statisticsName, statistics.maxValue());
        console.log("Min", statisticsName, statistics.minValue());
        console.log("Average", statisticsName, statistics.average());
        console.log("--------");
    }

    _initStatistic(id: string) {
        for (const statistics of [this._humidityStatistics, this._temperatureStatistics, this._pressureStatistics]) {
            if (!statistics.has(id)) {
                statistics.set(id, new Statistics());
            }
        }
    }

    private _temperatureStatistics: TSMap<string, Statistics> = new TSMap();
    private _humidityStatistics: TSMap<string, Statistics> = new TSMap();
    private _pressureStatistics: TSMap<string, Statistics> = new TSMap();
}

class WhetherData extends Observable<WhetherInfo> {
    constructor(name: string) {
        super();
        this._name = name;
    }

    public temperature(): number {
        return this._temperature;
    }

    public humidity(): number {
        return this._humidity;
    }

    public pressure(): number {
        return this._pressure;
    }

    public measurementsChanged() {
        this.notifyObservers();
    }

    public setMeasurements(temperature: number, humidity: number, pressure: number) {
        this._humidity = humidity;
        this._temperature = temperature;
        this._pressure = pressure;
        this.measurementsChanged();
    }

    protected _getObservableData() {
        return new WhetherInfo(this.temperature(), this.humidity(), this.pressure(), this._name);
    }

    private _name: string;
    private _temperature: number = 0.;
    private _humidity: number = 0.;
    private _pressure: number = 750.;
}

export {WhetherData};
export {WhetherInfo};
export {StatisticsDisplay};