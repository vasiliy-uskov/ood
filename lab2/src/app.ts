import {WhetherData} from "./WhetherData";
import {StatisticsDisplay} from "./WhetherData";

const wd: WhetherData = new WhetherData();
const display: StatisticsDisplay =  new StatisticsDisplay();
wd.addObserver(display);
wd.setMeasurements(3, 0.7, 760);
wd.setMeasurements(4, 0.8, 761);
wd.removeObserver(display);
wd.setMeasurements(10, 0.8, 761);
wd.setMeasurements(-10, 0.8, 761);