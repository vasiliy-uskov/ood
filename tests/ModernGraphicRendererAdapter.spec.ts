import {ModernGraphicRendererAdapter} from "../src/ModernGraphicRendererAdapter";

class MockOutputStream {
	print(str: string) {
		this.content = this.content + '\n' + str;
	}

	public content = '';
}

function creatAdapter(): {adapter: ModernGraphicRendererAdapter, mockStream: {readonly content: string}} {
	const mockStream = new MockOutputStream();
	const adapter = new ModernGraphicRendererAdapter(str => mockStream.print(str));
	return {
		mockStream,
		adapter,
	}
}

it('do not draw empty object, if call close without line', () => {
	const {mockStream, adapter} = creatAdapter();
	adapter.close();
	expect(mockStream.content).toMatchSnapshot();
});

it('can draw lines', () => {
	const {mockStream, adapter} = creatAdapter();
	adapter.moveTo(0, 0);
	adapter.lineTo(1, 1);
	adapter.close();
	expect(mockStream.content).toMatchSnapshot();
});

it('throw exception on lineTo call without start point', () => {
	const {adapter} = creatAdapter();
	expect(() => adapter.lineTo(1, 1)).toThrow();
});

it('start new draw object on moveTo', () => {
	const {mockStream, adapter} = creatAdapter();
	adapter.moveTo(0, 0);
	adapter.lineTo(1, 1);
	adapter.moveTo(0, 0);
	adapter.lineTo(1, 1);
	adapter.close();
	expect(mockStream.content).toMatchSnapshot();
});


it('start new draw object on close', () => {
	const {mockStream, adapter} = creatAdapter();
	adapter.moveTo(0, 0);
	adapter.lineTo(1, 1);
	adapter.close();
	adapter.moveTo(0, 0);
	adapter.lineTo(1, 1);
	expect(mockStream.content).toMatchSnapshot();
});

it('can set color for a lines', () => {
	const {mockStream, adapter} = creatAdapter();
	adapter.setColor(0xFF0000);
	adapter.moveTo(0, 0);
	adapter.lineTo(1, 1);
	adapter.close();
	expect(mockStream.content).toMatchSnapshot();
});