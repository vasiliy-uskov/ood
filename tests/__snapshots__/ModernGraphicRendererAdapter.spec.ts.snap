// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can draw lines 1`] = `
"
<draw>
	<line fromX=\\"0\\" fromY=\\"0\\" toX=\\"1\\" toY=\\"1\\">
		<color r=\\"0\\" g=\\"0\\" b=\\"0\\" a=\\"0\\" />
	</line>
</draw>"
`;

exports[`can set color for a lines 1`] = `
"
<draw>
	<line fromX=\\"0\\" fromY=\\"0\\" toX=\\"1\\" toY=\\"1\\">
		<color r=\\"1\\" g=\\"0\\" b=\\"0\\" a=\\"1\\" />
	</line>
</draw>"
`;

exports[`do not draw empty object, if call close without line 1`] = `""`;

exports[`start new draw object on close 1`] = `
"
<draw>
	<line fromX=\\"0\\" fromY=\\"0\\" toX=\\"1\\" toY=\\"1\\">
		<color r=\\"0\\" g=\\"0\\" b=\\"0\\" a=\\"0\\" />
	</line>
</draw>
<draw>
	<line fromX=\\"0\\" fromY=\\"0\\" toX=\\"1\\" toY=\\"1\\">
		<color r=\\"0\\" g=\\"0\\" b=\\"0\\" a=\\"0\\" />
	</line>"
`;

exports[`start new draw object on moveTo 1`] = `
"
<draw>
	<line fromX=\\"0\\" fromY=\\"0\\" toX=\\"1\\" toY=\\"1\\">
		<color r=\\"0\\" g=\\"0\\" b=\\"0\\" a=\\"0\\" />
	</line>
</draw>
<draw>
	<line fromX=\\"0\\" fromY=\\"0\\" toX=\\"1\\" toY=\\"1\\">
		<color r=\\"0\\" g=\\"0\\" b=\\"0\\" a=\\"0\\" />
	</line>
</draw>"
`;
